File: Codes in C++/Greedy Algorithm/Non Overlapping Intervals.cpp
class Solution {
public:
    static bool comp(vector<int>& a , vector<int>& b){
        if(a[1] == b[1]){
            return a[0]<b[0];
        }
        return a[1] < b[1];
    }
    int eraseOverlapIntervals(vector<vector<int>>& v) {
        sort(v.begin(),v.end(),comp);
        int res=0;
        int lastEnd=INT_MIN;
        for(int i=0;i<v.size();i++)
        {
            if(lastEnd>v[i][0])
            {
                res++;
            }
            else {
                lastEnd=v[i][1];
            }
        }
        return res;
    }
};

File: Codes in Python3/Intervals/Non-overlapping Intervals.py
#Sorting by Start
class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        count=0
        intervals.sort(key=lambda i:i[0])
        prevEnd=intervals[0][1]
        for start,end in intervals[1:]:
            if prevEnd>start:
                count+=1
                prevEnd=min(prevEnd,end)
            else:
                prevEnd=end

        return count

# TC=O(nlogn)
# SC=O(n)

#Sorting by End
class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        count=0
        intervals.sort(key=lambda i:i[1])
        prevEnd=intervals[0][1]
        for i in range(1, len(intervals)):
            if prevEnd>intervals[i][0]:
                count+=1
            else:
                prevEnd=intervals[i][1]

        return count

# TC=O(nlogn)
# SC=O(n)